File: Python/src/adt/bqs/bag_array.py
from __future__ import annotations

from collections.abc import Iterator
from random import Random

from .bag import Bag, Item
from .unordered_iterator import UnorderedIterator

#: The capacity a bag starts with. Growth doubles this, and only that doubling
#: goes through _grow_from.
INITIAL_CAPACITY = 32


class BagArray(Bag[Item]):
    """
    An implementation of the Bag interface using an array as the underlying storage.
    This class provides basic functionality such as adding elements, checking for
    containment, and retrieving the size of the bag. The bag does not maintain any
    specific order of elements. Internal capacity automatically grows when required
    to accommodate more items.

    NOTE: the backing store is deliberately a fixed-capacity list rather than a
    plain Python list. Growing it by hand is the point of the exercise in
    `_grow_from`; a list that grows itself would remove that exercise.
    """

    __slots__ = ("_items", "_count", "_random")

    def __init__(self, rnd: Random | None = None) -> None:
        """
        Construct an empty BagArray with an initial capacity of 32.

        Args:
            rnd: a random source, passed to any UnorderedIterator. If omitted, a
                 new Random is used. NOTE: a Random is mutable and therefore
                 unpredictable.
        """
        self._count: int = 0
        self._items: list[Item | None] | None = [None] * INITIAL_CAPACITY
        self._random: Random = rnd if rnd is not None else Random()

    @classmethod
    def of(cls, *items: Item, rnd: Random | None = None) -> BagArray[Item]:
        """
        Construct a BagArray containing the given items.

        NOTE this deliberately does not go through _grow_from: it allocates
        storage large enough at the outset. That matters because _grow_from is an
        exercise. A bag which could not even be constructed until the exercise was
        written made every test that merely uses a bag depend on it -- of the
        tests blocked that way, two thirds were testing graphs and classification
        sorts, not bags. Now only genuine growth, past the initial capacity, needs
        the exercise.

        The new bag has room to spare, so adding one more item does not
        immediately force a growth.

        Args:
            items: the items to put in the bag.
            rnd: a random source, as for the constructor.

        Returns:
            a BagArray containing exactly those items.
        """
        result = cls(rnd)
        if len(items) * 2 > INITIAL_CAPACITY:
            result._items = [None] * (len(items) * 2)
        assert result._items is not None
        result._items[: len(items)] = items
        result._count = len(items)
        return result

    def add(self, item: Item) -> None:
        """
        Add the specified item to the bag, expanding the internal storage if it is full.
        """
        assert self._items is not None
        if self._full():
            self._grow(self._items, 2 * self._capacity())
        self._items[self._count] = item
        self._count += 1

    def is_empty(self) -> bool:
        return self._count == 0

    def __len__(self) -> int:
        return self._count

    def clear(self) -> None:
        self._count = 0

    def contains(self, item: Item) -> bool:
        # NOTE the first _count entries only. Scanning the whole backing list
        # meant that after clear(), which resets the count and nothing else, this
        # still answered True for items the bag no longer held -- while
        # multiplicity, which happens to start with an is_empty() guard, answered
        # zero. Two methods disagreeing about the same question.
        assert self._items is not None
        for i in self._items[: self._count]:
            if i is not None and i == item:
                return True
        return False

    def __contains__(self, item: Item) -> bool:
        return self.contains(item)

    def multiplicity(self, item: Item) -> int:
        """
        Return the number of instances of item in this bag.
        """
        assert self._items is not None
        return sum(1 for i in self._items[: self._count] if i is not None and i == item)

    def as_array(self) -> list[Item]:
        """
        Return this bag's items as a list, excluding the unused tail of the
        backing store.
        """
        assert self._items is not None
        return list(self._items[: self._count])

    def __iter__(self) -> Iterator[Item]:
        """
        Return a randomly ordered iterator over this bag.
        """
        return UnorderedIterator(self.as_array(), self._random)

    def __repr__(self) -> str:
        return f"BagArray(items={self.as_array()!r}, count={self._count})"

    def _grow(self, source: list[Item | None], size: int) -> None:
        """
        Replace the internal storage with a copy of source expanded to size.
        """
        self._items = self._grow_from(source, size)

    def _capacity(self) -> int:
        assert self._items is not None
        return len(self._items)

    def _full(self) -> bool:
        return len(self) == self._capacity()

    @staticmethod
    def _grow_from(source: list[Item | None], size: int) -> list[Item | None]:
        """
        This fairly primitive grow method takes a list called source, creates a new
        list of the given size, copies all the elements of source into the start of
        the result, then returns the result. Unused slots must be None.

        Args:
            source: the source list.
            size: the size of the new list.
        """
        # TO BE IMPLEMENTED  grow array and copy
        raise NotImplementedError("TO BE IMPLEMENTED")

File: Java/src/main/java/com/phasmidsoftware/dsaipg/adt/bqs/Bag_Array.java
/*
 * Copyright (c) 2017. Phasmid Software
 */

package com.phasmidsoftware.dsaipg.adt.bqs;

import java.util.Arrays;
import java.util.Iterator;
import java.util.Random;

/**
 * An implementation of the Bag interface using an array as the underlying storage.
 * This class provides basic functionality such as adding elements, checking for containment,
 * and retrieving the size of the bag. The bag does not maintain any specific order of elements.
 * Internal capacity automatically grows when required to accommodate more items.
 *
 * @param <Item> the type of elements contained in the bag
 */
public class Bag_Array<Item> implements Bag<Item> {

    /**
     * Primary constructor that takes an explicit Random source (which will be passed to any UnorderedIterator).
     * NOTE: random is mutable and therefore unpredictable.
     *
     * @param random a random source.
     */
    public Bag_Array(Random random) {
        allocate(INITIAL_CAPACITY);
        this.random = random;
    }

    /**
     * Default no-argument constructor for Bag_Array.
     * Constructs a Bag_Array instance using a new Random instance as the source of entropy.
     * The constructor initializes the bag's internal storage with an initial capacity.
     */
    public Bag_Array() {
        this(new Random());
    }

    /**
     * Construct a Bag_Array containing the given items.
     * <p>
     * NOTE this deliberately does not call {@link #growFrom}: it allocates storage
     * large enough at the outset. That matters because growFrom is an exercise. A
     * bag which could not even be constructed until the exercise was written made
     * every test that merely uses a bag depend on it -- of the tests blocked that
     * way, two thirds were testing graphs and classification sorts, not bags.
     * Now only genuine growth, past the initial capacity, needs the exercise.
     * <p>
     * The new bag has room to spare, so adding one more item does not immediately
     * force a growth.
     *
     * @param items the items to put in the bag.
     * @param <T>   the item type.
     * @return a Bag_Array containing exactly those items.
     */
    @SafeVarargs
    public static <T> Bag_Array<T> of(T... items) {
        Bag_Array<T> result = new Bag_Array<>();
        if (items.length * 2 > INITIAL_CAPACITY) result.allocate(items.length * 2);
        System.arraycopy(items, 0, result.items, 0, items.length);
        result.count = items.length;
        return result;
    }

    /**
     * Adds the specified item to the bag.
     * If the internal storage is full, it will be expanded to accommodate the new item.
     *
     * @param item the item to be added to the bag
     */
    public void add(Item item) {
        assert items != null;
        if (full())
            grow(items, 2 * capacity());
        items[count++] = item;
    }

    /**
     * Determines if the bag currently contains no items.
     *
     * @return true if the bag is empty (i.e., the count of items is zero), otherwise false.
     */
    public boolean isEmpty() {
        return count == 0;
    }

    /**
     * Retrieves the number of items currently in the bag.
     *
     * @return the total count of items in the bag.
     */
    public int size() {
        return count;
    }

    /**
     * Clears all items from the bag.
     * Resets the count of items contained in the bag to zero, effectively emptying it.
     */
    public void clear() {
        count = 0;
    }

    /**
     * Checks if the specified item is present in the items array.
     *
     * @param item the item to check for presence in the bag
     * @return true if the item is found in the bag, otherwise false
     */
    public boolean contains(Item item) {
        // NOTE the first count entries only. Scanning the whole backing array
        // meant that after clear(), which resets count and nothing else, this
        // still answered true for items the bag no longer held -- while
        // multiplicity, which happens to start with an isEmpty() guard, answered
        // zero. Two methods disagreeing about the same question.
        for (int i = 0; i < count; i++) {
            Item x = items[i];
            if (x != null && x.equals(item))
                return true;
        }
        return false;
    }

    /**
     * Method to calculate the multiplicity of a given item in the bag.
     *
     * @param item the item for which the multiplicity is to be determined
     * @return the number of instances of the specified item in the bag
     */
    public int multiplicity(Item item) {
        int result = 0;
        for (int i = 0; i < count; i++) {
            Item x = items[i];
            if (x != null && x.equals(item))
                result++;
        }
        return result;
    }

    /**
     * Method to generate a randomly ordered iterator on this Bag.
     *
     * @return an Iterator on Item.
     */
    public Iterator<Item> iterator() {
        assert items != null; // Should be not-null any time after construction.
        return new UnorderedIterator<>(asArray(), random);
    }

    /**
     * Method to get this Bag as an array.
     * <p>
     * Internally, Object[] can be cast as an Item[] but it is not valid externally.
     * Hence, the arraycopy.
     * This is a quirk of Java Generics.
     *
     * @return this Bag as an array.
     */
    public Item[] asArray() {
        @SuppressWarnings("unchecked") Item[] items = (Item[]) new Object[count];
        System.arraycopy(this.items, 0, items, 0, count);
        return items;
    }

    @Override
    public String toString() {
        return "Bag_Array{" +
                "items=" + Arrays.toString(asArray()) +
                ", count=" + count +
                '}';
    }

    /**
     * Expands the Bag_Array's internal storage to accommodate a new size,
     * replacing the current items with the contents of the newly grown array.
     *
     * @param source the source array whose elements are to be copied into the new array
     * @param size   the size of the new expanded array
     */
    private void grow(Item[] source, int size) {
        items = growFrom(source, size);
    }

    /**
     * Allocate fresh, empty storage of the given size, discarding whatever was
     * there. Used at construction, where there is nothing to copy.
     *
     * @param size the size of the new array.
     */
    private void allocate(int size) {
        //noinspection unchecked
        items = (Item[]) new Object[size];
    }

    /**
     * Retrieves the current capacity of the internal storage for the bag.
     * The capacity is defined as the size of the array used to store items.
     * It represents the maximum number of items the bag can hold without resizing.
     *
     * NOTE package-private rather than private so that BagTest can assert that
     * growth DOUBLES the capacity. That property cannot be observed from outside:
     * an implementation which grew by a constant instead would satisfy every other
     * test, while making n additions cost O(n^2) rather than amortized O(n).
     *
     * @return the capacity of the internal storage.
     */
    int capacity() {
        assert items != null; // Should be not-null any time after construction.
        return items.length;
    }

    /**
     * Determines if the bag is full by checking if the number of items in the bag equals its capacity.
     *
     * @return true if the bag is full, otherwise false.
     */
    private boolean full() {
        return size() == capacity();
    }

    /**
     * This fairly primitive grow method takes a T array called "from",
     * instantiates a new array of the given size,
     * copies all the elements of from into the start of the resulting array,
     * then returns the result.
     *
     * NOTE package-private rather than private so that BagTest can test this
     * contract directly. Testing it only through {@code add} is not enough: add
     * always calls {@code grow(items, 2 * capacity())}, where the source is the
     * whole backing array, so an implementation returning an array of
     * {@code from.length * 2} is accidentally right at every call site while
     * ignoring the size it was given.
     *
     * @param from the source array
     * @param size the size of the new array
     */
    static <T> T[] growFrom(T[] from, int size) {
        // TO BE IMPLEMENTED  grow array and copy
                throw new com.phasmidsoftware.dsaipg.util.general.ImplementationMissing();
    }

    /**
     * The capacity a bag starts with. Growth doubles this, and only that doubling
     * goes through growFrom.
     */
    private static final int INITIAL_CAPACITY = 32;

    private final Random random;

    private Item[] items = null;
    private int count = 0;
}
